src/viam/sdk/common/proto_value.hpp

```diff
@@ -0,0 +1,324 @@
+        void swap(const vtable& this_vtable,
+                  storage& other,
+                  const vtable& other_vtable) noexcept(impl::all_proto_moves_noexcept{});
+
+        // Destroy this storage, using the destructor from vtable.
+        void destruct(const vtable& vtable) noexcept;
+
+        template <typename T = void>
+        T* get() {
+            return static_cast<T*>(static_cast<void*>(&buf_));
+        }
+
+        template <typename T = void>
+        T const* get() const {
+            return static_cast<T const*>(static_cast<void const*>(&buf_));
+        }
+
+        BufType buf_;
+    };
+
+    ProtoValue(const google::protobuf::Value* value);
+
+    vtable vtable_;
+    storage self_;
+};
+
+/// @brief Alias declaration for map of string to type-erased ProtoValue representing
+/// google::protobuf::Struct.
+/// This stores structured data as a map where the keys can be thought of as member names.
+using ProtoStruct = std::unordered_map<std::string, ProtoValue>;
+
+void to_proto(std::nullptr_t, google::protobuf::Value* v);
+void to_proto(bool b, google::protobuf::Value* v);
+void to_proto(int i, google::protobuf::Value* v);
+void to_proto(double d, google::protobuf::Value* v);
+void to_proto(std::string s, google::protobuf::Value* v);
+void to_proto(const std::vector<ProtoValue>& vec, google::protobuf::Value* v);
+void to_proto(const ProtoStruct& m, google::protobuf::Value* v);
+void to_proto(const ProtoValue& t, google::protobuf::Value* v);
+
+ProtoStruct struct_to_map(google::protobuf::Struct const* s);
+void map_to_struct(const ProtoStruct& m, google::protobuf::Struct* s);
+
+/// @brief Convert a type to a google::protobuf::Value.
+/// @note This method is trivially templated to insulate google::protobuf::Value from our
+/// API/ABI. It is meant to be called with no template parameters in a translation unit which
+/// includes <google/protobuf/struct.pb.h>
+template <typename T, typename Value = google::protobuf::Value>
+Value to_proto(T&& t) {
+    Value v;
+    to_proto(std::forward<T>(t), &v);
+
+    return v;
+}
+
+/// @brief Convert a google::protobuf::Struct to a ProtoStruct.
+/// @note This method is trivially templated to insulate google::protobuf::Struct from our
+/// API/ABI. It is meant to be called with no template parameters in a translation unit which
+/// includes <google/protobuf/struct.pb.h>
+template <typename Struct = google::protobuf::Struct>
+ProtoStruct struct_to_map(const Struct& s) {
+    return struct_to_map(&s);
+}
+
+/// @brief Convert a ProtoStruct to a google::protobuf::Struct.
+/// @note This method is trivially templated to insulate google::protobuf::Struct from our
+/// API/ABI. It is meant to be called with no template parameters in a translation unit which
+/// includes <google/protobuf/struct.pb.h>
+template <typename Struct = google::protobuf::Struct>
+Struct map_to_struct(const ProtoStruct& m) {
+    Struct s;
+    map_to_struct(m, &s);
+
+    return s;
+}
```

Comment:
Can we use this to work out what the general shape for all "SDK type" to "proto type" interconversions will look like? I suspect we want some sort of trait that binds the two together as well.

Reply:
Not sure if I follow, could you say more?

Reply:
For instance, in `board.hpp` we have these:

```
    /// @brief Creates a `power_mode` enum from its proto representation.
    static power_mode from_proto(viam::component::board::v1::PowerMode proto);

    /// @brief Converts a `power_mode` enum to its proto representation.
    static viam::component::board::v1::PowerMode to_proto(power_mode power_mode);
```

But I think we agree that we want to eliminate the "by size" usage of the proto types in our API, and move towards something more like what is being done in this review, where they are all defined in terms of opaque pointers:

```
static power_mode from_proto(viam::component::board::v1::PowerMode const* proto);
static to_proto(power_mode power_mode, viam::component::board::v1::PowerMode* proto);
```

But then we also want some templates to wrap this up more nicely so that if you have opted into having the `v1::PowerMode` type available you can use it with a better API:

```
v1::PowerMode to_proto(const power_mode& pm) {
    v1::PowerMode proto;
    to_proto(pm, &proto);
    return proto;
}

power_mode from_proto(const v1::PowerMode& proto) {
    return from_proto(&proto);
}
```

And I have some hand-wavy idea about whether there should be some sort of traits class which ties the two together:

```
template<typename ProtoType>
struct associated_sdk_type;

template<>
struct associated_sdk_type<v1::PowerMode> {
    using type = power_mode;
}

template<typename SdkType>
struct associated_proto_type;

template<>
struct associated_proto_type<power_mode> {
    using type = v1::PowerMode;
}
```

Because then maybe all the `to_proto` and `from_proto` things become one template? Or maybe you only need the trait for one direction because overloading handles the other?

But, generally, I'm interested in whether we can use this review to get the protocol sorted out that will be used globally in the SDK.

Reply:
Ahh I see, let me experiment a bit and I'll report back!